agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -1468,3 +1488,85 @@ func (dg *dockerGoClient) loadImage(ctx context.Context, reader io.Reader) error
 	}
 	return err
 }
+
+func getDevicesList(linkList []netlink.Link) []string {
+	var deviceNames []string
+	for _, link := range linkList {
+		if link.Type() != linkTypeDevice && link.Type() != linkTypeVlan {
+			// We only care about netlink.Device/netlink.Vlan types. Ignore other link types.
+			continue
+		}
+		if link.Attrs().EncapType == encapTypeLoopback {
+			// Ignore localhost
+			continue
+		}
+		deviceNames = append(deviceNames, link.Attrs().Name)
+	}
+	return deviceNames
+}
+
+func (dg *dockerGoClient) PopulateNIDeviceList(containerPID string) ([]string, error) {
+	var err error
+	var deviceList []string
+	nsClient := nswrapper.NewNS()
+	netlinkClient := netlinkwrapper.New()
+	netNSPath := fmt.Sprintf(ecscni.NetnsFormat, containerPID)
+	err = nsClient.WithNetNSPath(netNSPath, func(ns.NetNS) error {
+		linksInTaskNetNS, linkErr := netlinkClient.LinkList()
+		deviceNames := getDevicesList(linksInTaskNetNS)
+		deviceList = append(deviceList, deviceNames...)
+		return linkErr
+	})
+	return deviceList, err
+}
+
+func linkStatsToDockerStats(netLinkStats *netlink.LinkStatistics, numberOfContainers uint64) dockerstats.NetworkStats {
+	networkStats := dockerstats.NetworkStats{
+		RxBytes:   netLinkStats.RxBytes / numberOfContainers,
+		RxPackets: netLinkStats.RxPackets / numberOfContainers,
+		RxErrors:  netLinkStats.RxErrors / numberOfContainers,
+		RxDropped: netLinkStats.RxDropped / numberOfContainers,
+		TxBytes:   netLinkStats.TxBytes / numberOfContainers,
+		TxPackets: netLinkStats.TxPackets / numberOfContainers,
+		TxErrors:  netLinkStats.TxErrors / numberOfContainers,
+		TxDropped: netLinkStats.TxDropped / numberOfContainers,
+	}
+	return networkStats
+}
+
+func (dg *dockerGoClient) GetAWSVPCNetworkStats(deviceList []string, containerPID string,
```

Comment:
adding GetAWSVPCNetworkStats and PopulateNIDeviceList as new methods of docker client interface doesn't seem suitable to me because those stats are not related to docker. seems like they can just be helper function in agent/stats/task.go?

Reply:
I agree that these methods are not related to docker, I added them here for two reasons -- 

- Makes it easy to test. Because they are part of docker client interface it's easy to mock these function similar to how it is being done for 'stats' method -- https://github.com/aws/amazon-ecs-agent/blob/master/agent/dockerclient/dockerapi/docker_client.go#L167

- task.go is similar in structure to container.go. Container.go calls `stats` function to get the stats, it does not care about the implementation, going with the same pattern, task.go is calling `GetAWSVPCNetworkStats` and `PopulateNIDeviceList`. 

If it helps I can add more comments to these functions to make it clear what they are doing. Thoughts?

Reply:
for testing, you can make the nswrapper interface a field of the stats engine and a field of StatsTask, and then pass it from engine to StatsTask, and use mock for that field in test (similar to the way how docker client is mocked). for second point, IMO they don't have to be similar if the source of stats are essentially different. container stats are from a docker api call so it makes sense to use the docker client, but those network stats are from netlink.